source/distributed/fully_distributed_tria.cc

```diff
@@ -0,0 +1,534 @@
+          for (auto cell = this->begin(); cell != this->end(); ++cell)
+            {
+              if (cell->active())
+                cell->set_subdomain_id(
+                  dealii::numbers::artificial_subdomain_id);
+
+              cell->set_level_subdomain_id(
+                dealii::numbers::artificial_subdomain_id);
+            }
+
+          // 4b) set actual (level_)subdomain_ids as well as boundary ids
+          for (unsigned int level = 0; level < cell_infos.size(); ++level)
+            {
+              auto cell      = this->begin(level);
+              auto cell_info = cell_infos[level].begin();
+              for (; cell_info != cell_infos[level].end(); ++cell_info)
+                {
+                  // find cell that has the correct cell
+                  while (cell_info->id != cell->id().template to_binary<dim>())
+                    ++cell;
+
+                  // subdomain id
+                  if (cell->active())
+                    cell->set_subdomain_id(cell_info->subdomain_id);
+
+                  // level subdomain id
+                  if (settings & construct_multigrid_hierarchy)
+                    cell->set_level_subdomain_id(cell_info->level_subdomain_id);
+
+                  // boundary ids
+                  for (auto pair : cell_info->boundary_ids)
+                    {
+                      Assert(cell->at_boundary(pair.first),
+                             ExcMessage("Cell face is not on the boundary!"));
+                      cell->face(pair.first)->set_boundary_id(pair.second);
+                    }
+                }
+            }
+        }
+
+      update_number_cache();
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    Triangulation<dim, spacedim>::create_triangulation(
+      const std::vector<Point<spacedim>> &      vertices,
+      const std::vector<dealii::CellData<dim>> &cells,
+      const SubCellData &                       subcelldata)
+    {
+      AssertThrow(
+        currently_processing_create_triangulation_for_internal_usage,
+        ExcMessage(
+          "Use the other create_triangulation() function to create triangulations of type parallel::fullydistributed::Triangulation.!"));
+
+      dealii::Triangulation<dim, spacedim>::create_triangulation(vertices,
+                                                                 cells,
+                                                                 subcelldata);
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    Triangulation<dim, spacedim>::update_number_cache()
+    {
+      parallel::Triangulation<dim, spacedim>::update_number_cache();
+
+      if (settings & construct_multigrid_hierarchy)
+        parallel::Triangulation<dim, spacedim>::fill_level_ghost_owners();
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    Triangulation<dim, spacedim>::execute_coarsening_and_refinement()
+    {
+      Assert(false, ExcNotImplemented());
```

Comment:
The rule is that things that are programming errors only need to be checked in debug mode (`Assert`). This is one of those cases. On the other hand, things that depend on input files should always be checked (`AssertThrow`).